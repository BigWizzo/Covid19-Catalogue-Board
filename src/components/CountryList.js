import { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import getCountriesData from '../redux/actions/countriesActions';
import CountryListItem from './CountryListItem';

const reg = [
  'Search by Region',
  'Africa',
  'Europe',
  'Asia',
  'North America',
  'South America',
  'Oceania',
];

const CountryList = () => {
  const [nat, setNat] = useState('');
  const [region, setRegion] = useState('');

  const dispatch = useDispatch();
  const countries = useSelector((state) => state.countries.data);

  const fetchCountries = () => {
    dispatch(getCountriesData());
  };

  useEffect(() => {
    fetchCountries();
  }, []);

  const checkCountries = (byCountry) => {
    if (byCountry.All.country && byCountry.All.continent) {
      const selectByRegion = Object.values(byCountry.All.continent)
        .join('')
        .toLowerCase()
        .includes(region.toLowerCase());

      const selectByCountry = Object.values(byCountry.All.country)
        .join('')
        .toLowerCase()
        .includes(nat.toLowerCase());

      if (region.length > 0 && nat === '') {
        return selectByRegion;
      }
      if (nat.length > 0 && region === '') {
        return selectByCountry;
      }
      if (nat.length > 0 && region.length > 0) {
        return selectByCountry;
      }
    }
  };

  const filteredCountries = () => {
    if (nat === '' && region === '') {
      return countries;
    }
    return countries.filter((country) => checkCountries(country));
  };

  return (
    <div className="main">
      <h1 className="py-4 text-center">Covid19 Catalogue Board 1</h1>
      <form className="country-form m-auto mb-5">
        <div className="">
          <input
            className="form-control form-outline"
            type="text"
            value={nat}
            placeholder="Search by Country"
            onChange={(e) => setNat(e.target.value)}
          />
          <select
            className="form-select p-2"
            value={region}
            onChange={(e) => setRegion(e.target.value)}
            aria-label="Default select example"
          >
            {reg.map((x) => (
              <option value={x} key={x}>
                {x}
              </option>
            ))}
          </select>
        </div>
      </form>
      <div className="content">
        {countries.loading && <h4 className="text-center">Loading...</h4>}
        <div className="row g-0">
          {filteredCountries().map(
            (country) =>
              country.All.country && (
                <CountryListItem nation={country} key={country.All.country} />
              ),
          )}
        </div>
      </div>
      <div className="circle-small-4 br-50 pos-abs" />
      <div className="circle-small-5 br-50 pos-abs" />
      <div className="circle-small-2 br-50 pos-abs" />
    </div>
  );
};

export default CountryList;
