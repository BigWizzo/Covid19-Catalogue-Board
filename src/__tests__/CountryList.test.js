import React from 'react';
import renderer from 'react-test-renderer';
import CountryList from '../components/CountryList';
import { Provider } from 'react-redux';
import { store, persistor } from '../redux/store';
import { PersistGate } from 'redux-persist/integration/react';
import { BrowserRouter } from 'react-router-dom';

test('should render the list of all countries', () => {
  const component = renderer.create(
    <Provider store={store}>
      <PersistGate loading={null} persistor={persistor}>
        <BrowserRouter>
          <CountryList />
        </BrowserRouter>
      </PersistGate>
    </Provider>,
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
