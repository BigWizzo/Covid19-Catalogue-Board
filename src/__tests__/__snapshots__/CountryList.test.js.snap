// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the combined components to the App component 1`] = `
<div
  className="main"
>
  <h1
    className="py-4 text-center"
  >
    Covid 19 Board
  </h1>
  <form
    className="country-form m-auto mb-5"
  >
    <div
      className=""
    >
      <input
        className="form-control form-outline"
        onChange={[Function]}
        placeholder="Search by Country"
        type="text"
        value=""
      />
      <select
        aria-label="Default select example"
        className="form-select p-2"
        onChange={[Function]}
        value=""
      >
        <option
          value="Search by Region"
        >
          Search by Region
        </option>
        <option
          value="Africa"
        >
          Africa
        </option>
        <option
          value="Europe"
        >
          Europe
        </option>
        <option
          value="Asia"
        >
          Asia
        </option>
        <option
          value="North America"
        >
          North America
        </option>
        <option
          value="South America"
        >
          South America
        </option>
        <option
          value="Oceania"
        >
          Oceania
        </option>
      </select>
    </div>
  </form>
  <div
    className="content"
  >
    <div
      className="row g-0"
    />
  </div>
  <div
    className="circle-small-4 br-50 pos-abs"
  />
  <div
    className="circle-small-5 br-50 pos-abs"
  />
  <div
    className="circle-small-2 br-50 pos-abs"
  />
</div>
`;

exports[`should render the list of all countries 1`] = `
<div
  className="main"
>
  <h1
    className="py-4 text-center"
  >
    Covid 19 Board
  </h1>
  <form
    className="country-form m-auto mb-5"
  >
    <div
      className=""
    >
      <input
        className="form-control form-outline"
        onChange={[Function]}
        placeholder="Search by Country"
        type="text"
        value=""
      />
      <select
        aria-label="Default select example"
        className="form-select p-2"
        onChange={[Function]}
        value=""
      >
        <option
          value="Search by Region"
        >
          Search by Region
        </option>
        <option
          value="Africa"
        >
          Africa
        </option>
        <option
          value="Europe"
        >
          Europe
        </option>
        <option
          value="Asia"
        >
          Asia
        </option>
        <option
          value="North America"
        >
          North America
        </option>
        <option
          value="South America"
        >
          South America
        </option>
        <option
          value="Oceania"
        >
          Oceania
        </option>
      </select>
    </div>
  </form>
  <div
    className="content"
  >
    <div
      className="row g-0"
    />
  </div>
  <div
    className="circle-small-4 br-50 pos-abs"
  />
  <div
    className="circle-small-5 br-50 pos-abs"
  />
  <div
    className="circle-small-2 br-50 pos-abs"
  />
</div>
`;
